[BasicAA] Fix handling of indirect assumption based results

nikic opened the pull request:
If a result is potentially based on a not yet proven assumption,
BasicAA will remember it inside AssumptionBasedResults and remove
the cache entry it an assumption higher up is later disproven.
However, we currently miss the case where another cache entry ends
up depending on such an AssumptionBased result.

Fix this by introducing an additional AssumptionBased state for
cache entries. If such a result is used, we'll still increment
AAQI.NumAssumptionUses, which means that the using entry will
also become AssumptionBased and be cleared if the assumption is
disproven.

At the end of the root query, convert remaining AssumptionBased
results into definitive results.

Fixes https://github.com/llvm/llvm-project/issues/98978.

nikic committed:
[SLP] Precommit test for #98978 (NFC)

nikic commented on llvm/test/Transforms/SLPVectorizer/X86/pr98978.ll:
This test is a slightly reduced version from the original report in https://github.com/llvm/llvm-project/issues/98978. Unfortunately, the fact that this issue requires use of BatchAA makes it pretty hard to test, as one needs a specific interaction of IR and query order.

alexey-bataev replied:
The test and the changes looks good

fhahn commented on llvm/test/Transforms/SLPVectorizer/X86/pr98978.ll:
```diff
@@ -0,0 +1,127 @@
+  br label %loop1
+
+loop1:                                            ; preds = %bb20, %entry
+  %i = phi ptr [ %i21, %bb20 ], [ null, %entry ]
```
Might be good to avoid null pointers here and elsewhere if possible?

nikic replied:
I was able to get rid of one of the null pointers, but without the others the issue doesn't reproduce.

fhahn commented on llvm/test/Transforms/SLPVectorizer/X86/pr98978.ll:
```diff
@@ -0,0 +1,127 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -S -passes=slp-vectorizer < %s | FileCheck %s
```
Should we also have a BasicAA analysis test based on this?

nikic replied:
The BasicAA analysis tests don't use BatchAA. I think it might be possible to construct a case where this issue also occurs without BatchAA, but it's not easy.

fhahn replied:
Ah that's a shame, relying on SLP seems a bit fragile, but it sounds like it's the best option we have currently.

nikic committed:
[BasicAA] Fix handling of indirect assumption based results

If a result is potentially based on a not yet proven assumption,
BasicAA will remember it inside AssumptionBasedResults and remove
the cache entry it an assumption higher up is later disproven.
However, we currently miss the case where another cache entry ends
up depending on such an AssumptionBased result.

Fix this by introducing an additional AssumptionBased state for
cache entries. If such a result is used, we'll still increment
AAQI.NumAssumptionUses, which means that the using entry will
also become AssumptionBased and be cleared if the assumption is
disproven.

At the end of the root query, convert remaining AssumptionBased
results into definitive results.

Fixes https://github.com/llvm/llvm-project/issues/98978.

fhahn commented on llvm/test/Transforms/SLPVectorizer/X86/pr98978.ll:
```diff
@@ -0,0 +1,103 @@
+
+target triple = "x86_64-redhat-linux-gnu"
+
+; Should not get vectorized.
```
Would probably be helpful to include a bit more info about why it shouldn't get vectorized, as the test case isn't super easy to read.

nikic committed:
Expand comment

fhahn reviewed (approved):
LGTM, thanks!
